staging/src/k8s.io/api/certificates/v1/types.go

```diff
@@ -0,0 +1,226 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:method=UpdateApproval,verb=update,subresource=approval,input=k8s.io/api/certificates/v1.CertificateSigningRequest,result=k8s.io/api/certificates/v1.CertificateSigningRequest
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CertificateSigningRequest describes a certificate signing request
+type CertificateSigningRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// The certificate request itself and any additional information.
+	// +optional
+	Spec CertificateSigningRequestSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Derived information about the request.
+	// +optional
+	Status CertificateSigningRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CertificateSigningRequestSpec describes the requested certificate.
+// This information is immutable after the request is created.
+// Only the request, signerName, and usages fields can be set on creation,
+// other fields are derived by Kubernetes and cannot be modified by users.
+type CertificateSigningRequestSpec struct {
+	// request contains the base64-encoded PKCS#10 CSR data
+	// +listType=atomic
+	Request []byte `json:"request" protobuf:"bytes,1,opt,name=request"`
+
+	// signerName indicates the requested signer. It is a qualified name in the form:
+	// `scope-hostname.io/name`.
+	// Distribution of trust for signers happens out of band.
+	// You can select on this field using `spec.signerName`.
+	SignerName string `json:"signerName" protobuf:"bytes,7,opt,name=signerName"`
+
+	// usages specifies a set of key usages the issued certificate will be valid for.
+	// See: https://tools.ietf.org/html/rfc5280#section-4.2.1.3
+	//      https://tools.ietf.org/html/rfc5280#section-4.2.1.12
+	// Valid values are:
+	//  "signing",
+	//  "digital signature",
+	//  "content commitment",
+	//  "key encipherment",
+	//  "key agreement",
+	//  "data encipherment",
+	//  "cert sign",
+	//  "crl sign",
+	//  "encipher only",
+	//  "decipher only",
+	//  "any",
+	//  "server auth",
+	//  "client auth",
+	//  "code signing",
+	//  "email protection",
+	//  "s/mime",
+	//  "ipsec end system",
+	//  "ipsec tunnel",
+	//  "ipsec user",
+	//  "timestamping",
+	//  "ocsp signing",
+	//  "microsoft sgc",
+	//  "netscape sgc"
+	// +listType=atomic
+	Usages []KeyUsage `json:"usages,omitempty" protobuf:"bytes,5,opt,name=usages"`
+
+	// username contains the name of the requesting user.
```

Comment:
Is this filled out automatically?  If so I would say that, and I wouldn't reference user.Info interface because no one viewing this will see it.  Probably copy in.

Reply:
```go
// username contains the name of the user that created the CertificateSigningRequest.
// This field is populated by the API server on creationg and is immutable after creation.

// uid contains the uid of the user that created the CertificateSigningRequest.
// This field is populated by the API server on creationg and is immutable after creation.

// groups contains group membership of the user that created the CertificateSigningRequest.
// This field is populated by the API server on creationg and is immutable after creation.

// extra contains extra attributes of the user that created the CertificateSigningRequest.
// This field is populated by the API server on creationg and is immutable after creation.
```